File: mcp_server.py
"""
================================================================================
MacroQuant Gold Desk - Model Context Protocol (MCP) Server
================================================================================
Author: MacroQuant Fintech & AI Engineering
Target AI: Qwen 2.5 / 3.8 Max, Claude, GPT, Antigravity
Description:
    Standard stdio Model Context Protocol (MCP) Server enabling Qwen to:
    1. Read real-time Gold, DXY, and MT5 multi-timeframe confluence telemetry.
    2. Dynamically update & write JSON trading configurations, risk limits,
       affiliate URLs, King level thresholds, and EMA settings for the dashboard.
================================================================================
"""

import sys
import json
import os
import requests

# ------------------------------------------------------------------------------
# CONFIGURATION & FILE PATHS
# ------------------------------------------------------------------------------
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
CONFIG_FILE_PATH = os.path.join(BASE_DIR, "gold_config.json")
LIVE_FEED_FILE_PATH = os.path.join(BASE_DIR, "live_feed.json")
CLOUDFLARE_LIVE_URL = "https://gold.cjhomebase.fun/api/live"

DEFAULT_CONFIG = {
    "version": "1.0.0",
    "risk_management": {
        "max_risk_per_trade_percent": 1.5,
        "default_stop_loss_dollars": 10.0,
        "contract_size_oz": 100.0,
        "max_daily_drawdown_percent": 4.0
    },
    "technical_parameters": {
        "fast_ema_period": 8,
        "slow_ema_period": 21,
        "king_levels_lookback_bars": 50,
        "timeframe": "M15"
    },
    "affiliate_monetization": {
        "preferred_broker": "Vantage",
        "vantage_affiliate_url": "https://www.vantagemarkets.com/",
        "exness_affiliate_url": "https://www.exness.com/",
        "enable_cta_pulse_animations": True
    },
    "trading_rules": {
        "buy_dip_confluence": "H4/D1 == BULL and M15 == BEAR",
        "short_breakdown_confluence": "M15 == BEAR and H1 == BEAR and H4 == BEAR",
        "prohibit_trading_on_neutral": True
    }
}

def ensure_config_file():
    """Initializes the gold_config.json file if it does not exist."""
    if not os.path.exists(CONFIG_FILE_PATH):
        with open(CONFIG_FILE_PATH, "w", encoding="utf-8") as f:
            json.dump(DEFAULT_CONFIG, f, indent=2)

def read_config():
    """Reads the current gold_config.json configuration."""
    ensure_config_file()
    try:
        with open(CONFIG_FILE_PATH, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception as e:
        return {"error": f"Failed to read configuration: {str(e)}"}

def update_config(new_config_dict):
    """Deep merges or replaces configuration keys in gold_config.json."""
    ensure_config_file()
    try:
        current = read_config()
        if isinstance(new_config_dict, dict):
            current.update(new_config_dict)
            with open(CONFIG_FILE_PATH, "w", encoding="utf-8") as f:
                json.dump(current, f, indent=2)
            return {"status": "success", "updated_config": current}
        else:
            return {"error": "Invalid payload. Must be a JSON object."}
    except Exception as e:
        return {"error": f"Failed to update configuration: {str(e)}"}

def get_live_market_telemetry():
    """Fetches live Gold, DXY, and Confluence data from Cloudflare Worker or local cache."""
    try:
        resp = requests.get(CLOUDFLARE_LIVE_URL, timeout=2.0)
        if resp.status_code == 200:
            return resp.json()
    except Exception:
        pass

    # Fallback to local live_feed.json
    if os.path.exists(LIVE_FEED_FILE_PATH):
        try:
            with open(LIVE_FEED_FILE_PATH, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception:
            pass

    return {"status": "standby", "message": "Telemetry bridge initializing..."}

# ------------------------------------------------------------------------------
# MCP PROTOCOL RPC HANDLER (JSON-RPC 2.0 OVER STDIO)
# ------------------------------------------------------------------------------
TOOLS = [
    {
        "name": "get_gold_config",
        "description": "Retrieve the current JSON configuration parameters for the MacroQuant Gold Desk (Risk limits, EMA periods, King lookback, Affiliate URLs).",
        "inputSchema": {
            "type": "object",
            "properties": {},
            "required": []
        }
    },
    {
        "name": "update_gold_config",
        "description": "Update and modify the JSON trading settings, risk thresholds, EMA periods, or affiliate parameters for the Gold Desk.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "config": {
                    "type": "object",
                    "description": "JSON object containing updated settings (e.g. risk_management, technical_parameters, affiliate_monetization, trading_rules)."
                }
            },
            "required": ["config"]
        }
    },
    {
        "name": "get_gold_market_telemetry",
        "description": "Retrieve real-time live Gold price ticks, synthetic DXY index value, Multi-Timeframe EMA confluence (M15, H1, H4, D1), and 50-bar King structural levels.",
        "inputSchema": {
            "type": "object",
            "properties": {},
            "required": []
        }
    }
]

def handle_rpc_request(req):
    req_id = req.get("id")
    method = req.get("method")
    params = req.get("params", {})

    if method == "initialize":
        return {
            "jsonrpc": "2.0",
            "id": req_id,
            "result": {
                "protocolVersion": "2024-11-05",
                "capabilities": {
                    "tools": {}
                },
                "serverInfo": {
                    "name": "macroquant-gold-mcp",
                    "version": "1.0.0"
                }
            }
        }

    elif method == "tools/list":
        return {
            "jsonrpc": "2.0",
            "id": req_id,
            "result": {
                "tools": TOOLS
            }
        }

    elif method == "tools/call":
        tool_name = params.get("name")
        args = params.get("arguments", {})

        if tool_name == "get_gold_config":
            result_data = read_config()
            return {
                "jsonrpc": "2.0",
                "id": req_id,
                "result": {
                    "content": [{"type": "text", "text": json.dumps(result_data, indent=2)}]
                }
            }

        elif tool_name == "update_gold_config":
            new_conf = args.get("config", {})
            result_data = update_config(new_conf)
            return {
                "jsonrpc": "2.0",
                "id": req_id,
                "result": {
                    "content": [{"type": "text", "text": json.dumps(result_data, indent=2)}]
                }
            }

        elif tool_name == "get_gold_market_telemetry":
            telemetry = get_live_market_telemetry()
            return {
                "jsonrpc": "2.0",
                "id": req_id,
                "result": {
                    "content": [{"type": "text", "text": json.dumps(telemetry, indent=2)}]
                }
            }

        else:
            return {
                "jsonrpc": "2.0",
                "id": req_id,
                "error": {"code": -32601, "message": f"Tool '{tool_name}' not found."}
            }

    elif method == "notifications/initialized":
        return None

    else:
        return {
            "jsonrpc": "2.0",
            "id": req_id,
            "error": {"code": -32601, "message": f"Method '{method}' not found."}
        }

def main():
    ensure_config_file()
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            req = json.loads(line)
            res = handle_rpc_request(req)
            if res is not None:
                sys.stdout.write(json.dumps(res) + "\n")
                sys.stdout.flush()
        except Exception as e:
            err_res = {
                "jsonrpc": "2.0",
                "id": None,
                "error": {"code": -32700, "message": f"Parse error: {str(e)}"}
            }
            sys.stdout.write(json.dumps(err_res) + "\n")
            sys.stdout.flush()

if __name__ == "__main__":
    main()

File: mcp_server.js
#!/usr/bin/env node

/**
 * ============================================================================
 * MacroQuant Gold Desk - Node.js stdio MCP Server (Compatible with npx)
 * ============================================================================
 * Author: MacroQuant Fintech & AI Engineering
 * Target: Qwen 3.8 Max / Claude / Cursor / Antigravity via `npx`
 * ============================================================================
 */

const fs = require('fs');
const path = require('path');
const https = require('https');
const readline = require('readline');

// File paths
const CONFIG_FILE = path.join(__dirname, 'gold_config.json');
const LIVE_FEED_FILE = path.join(__dirname, 'live_feed.json');
const CLOUDFLARE_URL = 'https://gold.cjhomebase.fun/api/live';

const DEFAULT_CONFIG = {
  version: "1.0.0",
  risk_management: {
    max_risk_per_trade_percent: 1.5,
    default_stop_loss_dollars: 10.0,
    contract_size_oz: 100.0,
    max_daily_drawdown_percent: 4.0
  },
  technical_parameters: {
    fast_ema_period: 8,
    slow_ema_period: 21,
    king_levels_lookback_bars: 50,
    timeframe: "M15"
  },
  affiliate_monetization: {
    preferred_broker: "Vantage",
    vantage_affiliate_url: "https://www.vantagemarkets.com/",
    exness_affiliate_url: "https://www.exness.com/",
    enable_cta_pulse_animations: true
  },
  trading_rules: {
    buy_dip_confluence: "H4/D1 == BULL and M15 == BEAR",
    short_breakdown_confluence: "M15 == BEAR and H1 == BEAR and H4 == BEAR",
    prohibit_trading_on_neutral: true
  }
};

function ensureConfig() {
  if (!fs.existsSync(CONFIG_FILE)) {
    fs.writeFileSync(CONFIG_FILE, JSON.stringify(DEFAULT_CONFIG, null, 2), 'utf-8');
  }
}

function readConfig() {
  ensureConfig();
  try {
    const raw = fs.readFileSync(CONFIG_FILE, 'utf-8');
    return JSON.parse(raw);
  } catch (err) {
    return { error: `Failed to read config: ${err.message}` };
  }
}

function updateConfig(newConfig) {
  ensureConfig();
  try {
    let current = readConfig();
    if (typeof newConfig === 'object' && newConfig !== null) {
      current = { ...current, ...newConfig };
      fs.writeFileSync(CONFIG_FILE, JSON.stringify(current, null, 2), 'utf-8');
      return { status: "success", updated_config: current };
    }
    return { error: "Payload must be a JSON object" };
  } catch (err) {
    return { error: `Failed to update config: ${err.message}` };
  }
}

function fetchLiveTelemetry() {
  return new Promise((resolve) => {
    https.get(CLOUDFLARE_URL, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => {
        try {
          resolve(JSON.parse(data));
        } catch (e) {
          resolve(readLocalTelemetry());
        }
      });
    }).on('error', () => {
      resolve(readLocalTelemetry());
    });
  });
}

function readLocalTelemetry() {
  if (fs.existsSync(LIVE_FEED_FILE)) {
    try {
      return JSON.parse(fs.readFileSync(LIVE_FEED_FILE, 'utf-8'));
    } catch (e) {}
  }
  return { status: "standby", message: "Telemetry feed connecting..." };
}

// ----------------------------------------------------------------------------
// MCP PROTOCOL RPC DEFINITION
// ----------------------------------------------------------------------------
const TOOLS = [
  {
    name: "get_gold_config",
    description: "Retrieve active JSON configuration (Risk limits, EMA periods, King lookback, Affiliate URLs) for the Gold Desk terminal.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "update_gold_config",
    description: "Modify or inject new JSON trading parameters, risk settings, or affiliate links for the Gold Desk.",
    inputSchema: {
      type: "object",
      properties: {
        config: {
          type: "object",
          description: "JSON object containing updated settings."
        }
      },
      required: ["config"]
    }
  },
  {
    name: "get_gold_market_telemetry",
    description: "Get real-time Gold prices, synthetic DXY, Multi-Timeframe EMA confluence (M15, H1, H4, D1), and 50-bar King High/Low levels.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  }
];

async function handleRpc(req) {
  const { id, method, params } = req;

  if (method === "initialize") {
    return {
      jsonrpc: "2.0",
      id,
      result: {
        protocolVersion: "2024-11-05",
        capabilities: { tools: {} },
        serverInfo: { name: "macroquant-gold-mcp", version: "1.0.0" }
      }
    };
  }

  if (method === "tools/list") {
    return {
      jsonrpc: "2.0",
      id,
      result: { tools: TOOLS }
    };
  }

  if (method === "tools/call") {
    const { name, arguments: args } = params || {};

    if (name === "get_gold_config") {
      const cfg = readConfig();
      return {
        jsonrpc: "2.0",
        id,
        result: { content: [{ type: "text", text: JSON.stringify(cfg, null, 2) }] }
      };
    }

    if (name === "update_gold_config") {
      const res = updateConfig(args?.config || {});
      return {
        jsonrpc: "2.0",
        id,
        result: { content: [{ type: "text", text: JSON.stringify(res, null, 2) }] }
      };
    }

    if (name === "get_gold_market_telemetry") {
      const data = await fetchLiveTelemetry();
      return {
        jsonrpc: "2.0",
        id,
        result: { content: [{ type: "text", text: JSON.stringify(data, null, 2) }] }
      };
    }

    return {
      jsonrpc: "2.0",
      id,
      error: { code: -32601, message: `Tool '${name}' not found` }
    };
  }

  if (method === "notifications/initialized") return null;

  return {
    jsonrpc: "2.0",
    id,
    error: { code: -32601, message: `Method '${method}' not found` }
  };
}

// Stdio JSON-RPC interface
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

rl.on('line', async (line) => {
  if (!line.trim()) return;
  try {
    const req = JSON.parse(line);
    const res = await handleRpc(req);
    if (res !== null) {
      console.log(JSON.stringify(res));
    }
  } catch (err) {
    console.log(JSON.stringify({
      jsonrpc: "2.0",
      id: null,
      error: { code: -32700, message: `Parse error: ${err.message}` }
    }));
  }
});
